File: internal/sdk/marketplace.go
package sdk

// Marketplace resource (issue #54 CLI+SDK parity): the typed mirror of the
// agent-marketplace HTTP surface in cmd/api/marketplace.go.
//
// Endpoints (all under /v1/marketplace):
//
//	POST   /marketplace/listings            -> PublishListing  (OWNER/ADMIN)
//	GET    /marketplace/listings            -> BrowseListings  (any role; GLOBAL catalog)
//	GET    /marketplace/listings/{slug}     -> GetListing      (any role)
//	POST   /marketplace/listings/{slug}/install -> InstallListing (OWNER/ADMIN)
//	DELETE /marketplace/listings/{slug}     -> UnlistListing   (publisher org)
//
// Reads are global (published rows form the catalog for every org); writes
// are org-scoped from the credentials. Version snapshots contain config
// only — never request-supplied JSON — mirrored by ListingSnapshot.

import (
	"context"
	"net/url"
	"strconv"
	"strings"
	"time"
)

// ListingSnapshot is the immutable agent-config document embedded in a
// listing (name/description/instructions/model/status — config only).
type ListingSnapshot struct {
	Name         string `json:"name"`
	Description  string `json:"description"`
	Instructions string `json:"instructions"`
	Model        string `json:"model"`
	Status       string `json:"status"`
}

// Listing is one marketplace catalog entry.
type Listing struct {
	ID              string           `json:"id"`
	PublisherOrgID  string           `json:"publisher_org_id"`
	PublisherUserID string           `json:"publisher_user_id"`
	SourceAgentID   string           `json:"source_agent_id"`
	Name            string           `json:"name"`
	Slug            string           `json:"slug"`
	Description     string           `json:"description"`
	Tags            []string         `json:"tags"`
	Status          string           `json:"status"` // draft|published|unlisted
	DownloadCount   int              `json:"download_count"`
	VersionSnapshot *ListingSnapshot `json:"version_snapshot"`
	CreatedAt       time.Time        `json:"created_at"`
	UpdatedAt       time.Time        `json:"updated_at"`
}

// ListingPage is the wrapped shape of GET /v1/marketplace/listings with its
// keyset-pagination cursor (NextCursor is "" when the page is exhausted).
type ListingPage struct {
	Listings   []Listing `json:"listings"`
	NextCursor string    `json:"next_cursor"`
}

// BrowseOptions selects the catalog search. Query is a case-insensitive
// substring match over name/description; Tags is an ANY-overlap filter; the
// page size and cursor mirror the server's keyset pagination.
type BrowseOptions struct {
	Query  string
	Tags   []string
	Limit  int // <= 0 lets the server default
	Cursor string
}

// PublishListingRequest is the POST /v1/marketplace/listings body. Version 0
// (the default) snapshots the agent's CURRENT live configuration; a positive
// version publishes that immutable config-version snapshot. Slug and
// description are optional (derived server-side when empty). Status is
// "published" (default) or "draft".
type PublishListingRequest struct {
	AgentID     string   `json:"agent_id"`
	Version     int      `json:"version,omitempty"`
	Name        string   `json:"name,omitempty"`
	Slug        string   `json:"slug,omitempty"`
	Description string   `json:"description,omitempty"`
	Tags        []string `json:"tags,omitempty"`
	Status      string   `json:"status,omitempty"`
}

// InstalledAgent is the newly created agent of an install response
// (snake_case — distinct from the legacy PascalCase Agent wire shape the
// /agents handlers emit).
type InstalledAgent struct {
	ID             string    `json:"id"`
	OrganizationID string    `json:"organization_id"`
	Name           string    `json:"name"`
	Description    string    `json:"description"`
	Instructions   string    `json:"instructions"`
	Model          string    `json:"model"`
	Status         string    `json:"status"`
	CreatedAt      time.Time `json:"created_at"`
}

// InstallResult is the 201 body of POST /v1/marketplace/listings/{slug}/install:
// the listing (with its refreshed download_count) plus the freshly created
// agent in the caller's org. Name collisions get a deterministic -2/-3/…
// suffix server-side.
type InstallResult struct {
	Listing Listing        `json:"listing"`
	Agent   InstalledAgent `json:"agent"`
}

// BrowseListings searches the GLOBAL published catalog
// (GET /v1/marketplace/listings?q&tags&limit&cursor).
func (c *Client) BrowseListings(ctx context.Context, opts BrowseOptions) (*ListingPage, error) {
	query := make(url.Values)
	if opts.Query != "" {
		query.Set("q", opts.Query)
	}
	if len(opts.Tags) > 0 {
		query.Set("tags", strings.Join(opts.Tags, ","))
	}
	if opts.Limit > 0 {
		query.Set("limit", strconv.Itoa(opts.Limit))
	}
	if opts.Cursor != "" {
		query.Set("cursor", opts.Cursor)
	}
	var out ListingPage
	if err := c.do(ctx, httpMethodGet, "/marketplace/listings", query, nil, &out); err != nil {
		return nil, err
	}
	if out.Listings == nil {
		out.Listings = []Listing{}
	}
	return &out, nil
}

// GetListing resolves one listing by its globally-unique slug
// (GET /v1/marketplace/listings/{slug}). Draft/unlisted listings are visible
// only to their publisher org (foreign callers get 404).
func (c *Client) GetListing(ctx context.Context, slug string) (*Listing, error) {
	var out struct {
		Listing Listing `json:"listing"`
	}
	if err := c.do(ctx, httpMethodGet, "/marketplace/listings/"+urlPathEscape(slug), nil, nil, &out); err != nil {
		return nil, err
	}
	return &out.Listing, nil
}

// PublishListing turns an existing agent of the caller's org into a catalog
// listing (POST /v1/marketplace/listings, 201).
func (c *Client) PublishListing(ctx context.Context, req PublishListingRequest) (*Listing, error) {
	var out struct {
		Listing Listing `json:"listing"`
	}
	if err := c.do(ctx, httpMethodPost, "/marketplace/listings", nil, req, &out); err != nil {
		return nil, err
	}
	return &out.Listing, nil
}

// InstallListing creates a NEW agent in the caller's org from the listing
// snapshot and bumps the download counter
// (POST /v1/marketplace/listings/{slug}/install, 201).
func (c *Client) InstallListing(ctx context.Context, slug string) (*InstallResult, error) {
	var out InstallResult
	if err := c.do(ctx, httpMethodPost, "/marketplace/listings/"+urlPathEscape(slug)+"/install", nil, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// UnlistListing removes a listing from the public catalog
// (DELETE /v1/marketplace/listings/{slug}). The caller's org must BE the
// publisher org; foreign/unknown slugs surface as 404.
func (c *Client) UnlistListing(ctx context.Context, slug string) error {
	var out struct {
		Deleted bool `json:"deleted"`
	}
	return c.do(ctx, httpMethodDelete, "/marketplace/listings/"+urlPathEscape(slug), nil, nil, &out)
}

File: web/src/lib/api/marketplace.ts
// Marketplace resource (issue #28, cmd/api/marketplace.go):
//
//   GET    /marketplace/listings             -> {"listings":[…],"next_cursor":…}  browse (agents.read — any role)
//   POST   /marketplace/listings             -> {"listing":{…}}                    publish (agents.write — OWNER/ADMIN)
//   GET    /marketplace/listings/{slug}      -> {"listing":{…}}                    get (agents.read — any role)
//   POST   /marketplace/listings/{slug}/install -> {"listing":{…},"agent":{…}}     install (agents.write — OWNER/ADMIN)
//   DELETE /marketplace/listings/{slug}      -> {"deleted":true}                   unlist (agents.write — OWNER/ADMIN)
//
// Browse is a GLOBAL read (the published catalog for every organization);
// publish/install are org-scoped writes derived from the auth claims. Browse
// is keyset-paginated: pass `cursor` from the previous response's next_cursor.

import { apiFetch } from './client'
import { asNumber, asString, pickField } from './types'
import { normalizeAgent, type Agent } from './types'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type MarketplaceListing = {
  id: string
  publisherOrgId: string
  publisherUserId: string
  sourceAgentId: string
  name: string
  slug: string
  description: string
  tags: string[]
  status: string
  downloadCount: number
  /** Decoded version_snapshot (agent config only — never secrets). */
  versionSnapshot: Record<string, unknown> | null
  createdAt?: string
  updatedAt?: string
}

export type BrowseListingsInput = {
  query?: string
  tags?: string[]
  limit?: number
  cursor?: string
}

export type BrowseListingsResult = {
  listings: MarketplaceListing[]
  nextCursor: string
}

export type InstallListingResult = {
  listing: MarketplaceListing
  agent: Agent
}

export type PublishListingInput = {
  agent_id: string
  /** >0 publishes the immutable config version with that number; 0/omitted snapshots the live config. */
  version?: number
  name: string
  slug?: string
  description?: string
  tags?: string[]
  /** "published" (default) or "draft". */
  status?: string
}

/** Backend slug regex (internal/marketplace/service.go), for form pre-validation. */
export const MARKETPLACE_SLUG_PATTERN = /^[a-z0-9]([a-z0-9-]*[a-z0-9])?$/

// ---------------------------------------------------------------------------
// Normalizers
// ---------------------------------------------------------------------------

function normalizeListing(raw: unknown): MarketplaceListing {
  const snapshot = pickField(raw, 'versionSnapshot', 'version_snapshot')
  const tags = pickField(raw, 'tags')
  return {
    id: asString(pickField(raw, 'id')) ?? '',
    publisherOrgId: asString(pickField(raw, 'publisherOrgId', 'publisher_org_id')) ?? '',
    publisherUserId: asString(pickField(raw, 'publisherUserId', 'publisher_user_id')) ?? '',
    sourceAgentId: asString(pickField(raw, 'sourceAgentId', 'source_agent_id')) ?? '',
    name: asString(pickField(raw, 'name')) ?? 'Unnamed listing',
    slug: asString(pickField(raw, 'slug')) ?? '',
    description: asString(pickField(raw, 'description')) ?? '',
    tags: Array.isArray(tags) ? tags.map((tag) => asString(tag) ?? '').filter((tag) => tag !== '') : [],
    status: (asString(pickField(raw, 'status')) ?? 'unknown').toLowerCase(),
    downloadCount: asNumber(pickField(raw, 'downloadCount', 'download_count')) ?? 0,
    versionSnapshot: snapshot && typeof snapshot === 'object' && !Array.isArray(snapshot)
      ? { ...(snapshot as Record<string, unknown>) }
      : null,
    createdAt: asString(pickField(raw, 'createdAt', 'created_at')),
    updatedAt: asString(pickField(raw, 'updatedAt', 'updated_at')),
  }
}

// ---------------------------------------------------------------------------
// Fetchers
// ---------------------------------------------------------------------------

export async function browseListings(input: BrowseListingsInput = {}): Promise<BrowseListingsResult> {
  const params = new URLSearchParams()
  if (input.query) params.set('q', input.query)
  if (input.tags && input.tags.length > 0) params.set('tags', input.tags.join(','))
  if (input.limit && input.limit > 0) params.set('limit', String(input.limit))
  if (input.cursor) params.set('cursor', input.cursor)
  const query = params.toString()
  const raw = await apiFetch<unknown>(`/marketplace/listings${query ? `?${query}` : ''}`)
  const list = pickField(raw, 'listings')
  return {
    listings: (Array.isArray(list) ? list : []).map(normalizeListing),
    nextCursor: asString(pickField(raw, 'nextCursor', 'next_cursor')) ?? '',
  }
}

export async function installListing(slug: string): Promise<InstallListingResult> {
  const raw = await apiFetch<unknown>(`/marketplace/listings/${encodeURIComponent(slug)}/install`, {
    method: 'POST',
  })
  return {
    listing: normalizeListing(pickField(raw, 'listing') ?? raw),
    agent: normalizeAgent(pickField(raw, 'agent')),
  }
}

export async function publishListing(input: PublishListingInput): Promise<MarketplaceListing> {
  const raw = await apiFetch<unknown>('/marketplace/listings', {
    method: 'POST',
    body: JSON.stringify({
      agent_id: input.agent_id,
      version: input.version ?? 0,
      name: input.name,
      slug: input.slug ?? '',
      description: input.description ?? '',
      tags: input.tags ?? [],
      status: input.status ?? 'published',
    }),
  })
  return normalizeListing(pickField(raw, 'listing') ?? raw)
}
